File: scripts_go/CPCTF_Bitwise_Scrumble.go
package main

import (
	"fmt"
	"math/big"
	"strconv"
)

// CPCTF - Bitwise Scrumble
//
// 附件把 flag 转成 75 位十进制整数串后分成 3 个 25 位块。
// 每一位十进制数字都会和对应 key 数字做看似不同的 bitwise 表达式。
// 根据真值表可化简出三种表达式都等价于 digit ^ key_digit，
// 因此解密时再异或同一个 key_digit 即可还原原十进制数字。

const (
	key = "0123456789012109876543210"
	enc = "10aa77170b38758c146245779086332e5e8237430f362d317310124333b999b890043152135"
)

func recoverDecimalDigits(part string) string {
	result := make([]byte, 0, len(part))
	for i := 0; i < len(part); i++ {
		keyDigit, err := strconv.Atoi(string(key[i]))
		if err != nil {
			panic(err)
		}
		encryptedNibble, err := strconv.ParseInt(string(part[i]), 16, 64)
		if err != nil {
			panic(err)
		}
		result = append(result, byte(int(encryptedNibble)^keyDigit)+'0')
	}
	return string(result)
}

func main() {
	firstPart := enc[:25]
	secondPart := enc[25:50]
	thirdPart := enc[50:75]

	decimalText := recoverDecimalDigits(firstPart) +
		recoverDecimalDigits(secondPart) +
		recoverDecimalDigits(thirdPart)

	flagNumber := new(big.Int)
	if _, ok := flagNumber.SetString(decimalText, 10); !ok {
		panic("failed to parse recovered decimal text")
	}

	fmt.Println(string(flagNumber.Bytes()))
}

File: scripts_python/CPCTF_Bitwise_Scrumble.py
"""
CPCTF - Bitwise Scrumble

每个十进制数字与 key 数字独立进行 bitwise 运算。写出单 bit 真值表后可知，
附件中的三种表达式都等价于 `digit ^ key_digit`，所以密文 hex nibble 再异或
同一个 key digit 就能还原原始十进制数字串。
"""

from __future__ import annotations


KEY = "0123456789012109876543210"
ENCRYPTED_FLAG = "10aa77170b38758c146245779086332e5e8237430f362d317310124333b999b890043152135"


def long_to_bytes(value: int) -> bytes:
    return value.to_bytes((value.bit_length() + 7) // 8, "big")


def recover_decimal_digits(part: str) -> str:
    digits = []
    for encrypted_nibble, key_digit in zip(part, KEY):
        digits.append(str(int(encrypted_nibble, 16) ^ int(key_digit)))
    return "".join(digits)


def solve() -> str:
    parts = (
        ENCRYPTED_FLAG[:25],
        ENCRYPTED_FLAG[25:50],
        ENCRYPTED_FLAG[50:75],
    )
    decimal_text = "".join(recover_decimal_digits(part) for part in parts)
    return long_to_bytes(int(decimal_text)).decode()


def main() -> None:
    print(solve())


if __name__ == "__main__":
    main()
